server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java

```diff
@@ -49,6 +49,15 @@
 
     ShardId shardId();
 
+    /**
+     * Returns the remapped shard id of the requested shard for this request
+     * or -1 if this information is not available.
+     * The remapped shard id is the id of the requested shard among all shards
+     * of this index that are part of the request. Note that the remapped shard id
+     * is equal to the original shard id if all shards of this index are part of the request.
+     */
+    int remapShardId();
```

Comment:
I wonder if we should call it `shardRequestOrdinal` shard ID is trappy at least for me since I alwasy think of ShardID.java when I read it.

Reply:
+1. I pushed https://github.com/elastic/elasticsearch/pull/29533/commits/b90716cfa7eb17b10de0d8398220431fbe856da5